linkerd/app/outbound/src/http/concrete.rs

```diff
@@ -1,104 +1,368 @@
-use super::{Concrete, Endpoint};
-use crate::{endpoint, stack_labels, Outbound};
+//! A stack that (optionally) resolves a service to a set of endpoint replicas
+//! and distributes HTTP requests among them.
+
+use super::{balance, client, normalize_uri};
+use crate::{http, stack_labels, Outbound};
 use linkerd_app_core::{
+    metrics, profiles,
     proxy::{
-        api_resolve::{ConcreteAddr, Metadata},
+        api_resolve::{ConcreteAddr, Metadata, ProtocolHint},
         core::Resolve,
-        http,
+        tap,
     },
-    svc, Error,
+    svc::{self, Layer},
+    tls,
+    transport::{self, addrs::*},
+    Error, Infallible, NameAddr,
 };
-use std::time;
+use std::{fmt::Debug, net::SocketAddr};
 use tracing::info_span;
 
+/// Parameter configuring dispatcher behavior.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub enum Dispatch {
+    Balance(NameAddr, balance::EwmaConfig),
+    Forward(Remote<ServerAddr>, Metadata),
+}
+
+/// Wraps errors encountered in this module.
 #[derive(Debug, thiserror::Error)]
 #[error("concrete service {addr}: {source}")]
 pub struct ConcreteError {
-    addr: ConcreteAddr,
+    addr: NameAddr,
     #[source]
     source: Error,
 }
 
+/// Inner stack target type.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Endpoint<T> {
+    addr: Remote<ServerAddr>,
+    is_local: bool,
+    metadata: Metadata,
+    parent: T,
+}
+
+/// A target configuring a load balancer stack.
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct Balance<T> {
+    addr: NameAddr,
+    ewma: balance::EwmaConfig,
+    parent: T,
+}
+
+// === impl Outbound ===
+
 impl<N> Outbound<N> {
     /// Builds a [`svc::NewService`] stack that builds buffered HTTP load
-    /// balancer services for [`Concrete`] targets.
+    /// balancer services for `T`-typed concrete targets.
     ///
     /// When a balancer has no available inner services, it goes into
     /// 'failfast'. While in failfast, buffered requests are failed and the
     /// service becomes unavailable so callers may choose alternate concrete
     /// services.
```

Comment:
should this comment maybe be updated to indicate that this stack also now determines whether to build a load balancer or forward directly to an endpoint?